### PR TITLE
Add simpler permit2 wrapper in `Permit2Lib`

### PR DESCRIPTION
**Pull request by MerlinEgalite**
## Motivation

`Permit2`'s goal is to greatly improve UX for EOAs. The dev experience can also be improved by easing `Permit2`'s integration with other protocols (and thus increasing its network effect). This, I guess, is one of the goals of the `Permit2Lib`. 

To further improve the dev exp, I propose to remove parts of the `permit2` function in the `Permit2Lib` that are specific to some tokens. Then, integrators just need to implement approval on the `Permit2` contract without worrying about whether the token already implements (and how) the `permit` functionality or not.

Of course, an account can still directly call the token to give its allowance to an arbitrary address. That's why the `transferFrom2` function hasn’t been modified.

I understand that this proposition is opinionated.

### CODE REVIEW

**Comment by transmissions11** [comment]
handling these edge cases is basically the whole point of the lib haha but i understand the desire for a simple wrapper over permit2 with the same interface as permit. 

happy to compromise and add a `simplePermit2` function or something to the lib for your use case

**Comment by MerlinEgalite** [comment]
> handling these edge cases is basically the whole point of the lib haha but i understand the desire for a simple wrapper over permit2 with the same interface as permit.
> 
> happy to compromise and add a `simplePermit2` function or something to the lib for your use case— but i'm strongly against just cutting out whats in there

Haha yes, I agree. That's more to showcase how we would use it. I'll update the PR to create this function. I still think this simple function is useful as it abstracts away the complexity of preparing the payload to call the `permit` function on the `Permit2` contract.

File: src/libraries/Permit2Lib.sol [hunk 1]
```diff
@@ -140,4 +141,44 @@ library Permit2Lib {
...
+    /// @notice Permit a user to spend a given amount of
+    /// another user's tokens via the owner's EIP-712 signature.
+    /// @param token The token to permit spending.
+    /// @param owner The user to permit spending from.
+    /// @param spender The user to permit spending to.
+    /// @param amount The amount to permit spending.
+    /// @param deadline  The timestamp after which the signature is no longer valid.
+    /// @param v Must produce valid secp256k1 signature from the owner along with r and s.
+    /// @param r Must produce valid secp256k1 signature from the owner along with v and s.
+    /// @param s Must produce valid secp256k1 signature from the owner along with r and v.
```
Comment (transmissions11) [inline src/libraries/Permit2Lib.sol:154]: nice this looks good! can we maybe add a note about the fact that it uses an ~unlimited expiration of `type(uint48).max`? not necessarily obvious

Comment (transmissions11) [reply src/libraries/Permit2Lib.sol:154]: could also re-use this function in `permit2` maybe? if it adds less than 100 gas might be worth it for cleanliness

Comment (MerlinEgalite) [reply src/libraries/Permit2Lib.sol:154]: There's already this note
https://github.com/MerlinEgalite/permit2/blob/754063e8eaaf184e94160520aa7febd272fb43a6/src/libraries/Permit2Lib.sol#L173-L174

Should we duplicate it? 

Way cleaner you're right!

Comment (transmissions11) [reply src/libraries/Permit2Lib.sol:135]: > Should we duplicate it?

yeah i meant as a doc comment, that way ppl hovering over the function in their IDEs, etc see it

File: src/libraries/Permit2Lib.sol [hunk 2]
```diff
@@ -119,25 +120,46 @@ library Permit2Lib {
...
-                        token: address(token),
-                        amount: amount.toUint160(),
-                        // Use an unlimited expiration because it most
-                        // closely mimics how a standard approval works.
-                        expiration: type(uint48).max,
-                        nonce: nonce
-                    }),
-                    spender: spender,
-                    sigDeadline: deadline
+    /// @notice Permit on the Permit2 contract.
```
Comment (transmissions11) [inline src/libraries/Permit2Lib.sol:127]:
```suggestion
    /// @notice Simple unlimited permit on the Permit2 contract.
```

File: src/libraries/Permit2Lib.sol [hunk 3]
```diff
@@ -58,7 +58,8 @@ library Permit2Lib {
     //////////////////////////////////////////////////////////////*/
 
     /// @notice Permit a user to spend a given amount of
-    /// another user's tokens via the owner's EIP-712 signature.
+    /// another user's tokens via the owner's EIP-712 signature
+    /// trying to call the native permit function of the token first (if any).
```
Comment (transmissions11) [inline src/libraries/Permit2Lib.sol:62]:
```suggestion
    /// another user's tokens via native EIP-2612 permit if possible falling
    /// back to Permit2 if native permit fails or is not implemented on the token.
```

**Review by transmissions11 — APPROVED** [review]

File: src/libraries/Permit2Lib.sol (hunk 1 above)
Comment (snreynolds) [reply src/libraries/Permit2Lib.sol:135]: Hmm don't see the expiration comment. Think it would be good to add as an @dev tag at the top.

**Pull request closed without merging**